src/main.go

```diff
@@ -91,13 +95,39 @@ func addKeys(akd *keys.SSHAuthorizedKeysDir, name string, force bool) error {
 		k = append(k, b...)
 	}
 
-	// TODO(vc): akd.Add() treats keys as opaque objects, so we need to
-	// sanity check them ourselves before adding them.
-	// update-ssh-keys employs ssh-keygen -l -f for this purpose, can do
-	// same here.
+	if !validKeys(k) {
+		return fmt.Errorf("key(s) invalid.")
+	}
+
 	return akd.Add(name, k, *flagReplace, force)
 }
 
+// validKeys validates a byte slice contains valid ssh keys.
+func validKeys(keys []byte) bool {
+	// we need to write out a temp file for ssh-keygen to consume.
+	tf, err := ioutil.TempFile("", "update-ssh-keys-")
+	if err != nil {
+		panicf("unable to create temporary file: %v", err)
+	}
+	defer func() {
+		os.Remove(tf.Name())
+		tf.Close()
+	}()
+
+	if _, err := tf.Write(keys); err != nil {
+		panicf("unable to write temporary file: %v", err)
+	}
+
+	cmd := exec.Command("ssh-keygen", "-l", "-f", tf.Name())
```

Comment:
Although this is equivalent with the existing update-ssh-keys script it is actually incomplete. `ssh-keygen -l` only reads the first key in the file since it assumes it is operating on an identity like `id_rsa`/`id_rsa.pub` instead of a list of keys like `authorized_keys`. This is fine for now but a simple but more complete solution would probably be to pair http://golang.org/pkg/bufio/#Scanner and https://godoc.org/golang.org/x/crypto/ssh#ParseAuthorizedKey